drivers: bluetooth: fix and clean up BlueNRG HCI setup() function

arbrauns opened the pull request:
If no public address was set, it should just do nothing instead of erroring. The function should also be static and there's no need to copy the address struct.

Ref #65179

arbrauns committed:
drivers: bluetooth: fix and clean up BlueNRG HCI setup() function

If no public address was set, it should just do nothing instead of
erroring. The function should also be static and there's no need to copy
the address struct.

Signed-off-by: Armin Brauns <armin.brauns@embedded-solutions.at>

msmttchr reviewed (commented):
@arbrauns, thanks for the quick fix, I left some minor comment from my side.
I have few other points to highlight:
- function `bt_spi_bluenrg_setup` is included even if users do not need to perform the setup, increasing (very slightly) the code size
- The field `setup` is initialized conditionally to `CONFIG_BT_BLUENRG_ACI`, while the `struct bt_hci_driver` is declared in `hci_driver.h` and the field `setup` is conditionally declared under the control of `CONFIG_BT_HCI_SETUP`. I know the two CONFIG_* are kept in sync by `Kconfig` definition, but this is not guaranteed for the future and so you can have compilation issues if the two CONFIG get misaligned, so I would put field `setup`  under `CONFIG_BT_HCI_SETUP`.
- I would leave to the user to select whether he wants to enable the flexibility to add the public address (and include the relevant code) . Looking at https://github.com/zephyrproject-rtos/zephyr/pull/65179#discussion_r1410570327, I guess you do not agree, but just to have your comment.

Do you think it is relevant to add some documentation on how to use this feature ? My understanding is:

1. Call `bt_id_create` to create an identity with a public address
2. Call `bt_enable`

Few other questions:
- Which platform did you to test these changes ? I can offer to test some ST platform as well
- May be not for you: if I take a platform like [96_carbon](https://docs.zephyrproject.org/latest/boards/arm/96b_carbon/doc/index.html ) based on nRF51822 controller, how would you set the public address on it ?

Thanks again and sorry for the long comment.

jori-nordic commented:
> but I am not sure whether it’s a suggested practice in Zephyr

The style in the host is rather explicit `return -ERRx` instead of single point.

arbrauns commented on drivers/bluetooth/hci/spi.c:
```diff
@@ -243,22 +243,20 @@ static int bt_spi_send_aci_config(uint8_t offset, const uint8_t *value, size_t v
-		return ret;
+		if (ret != 0) {
+			LOG_ERR("Failed to set BlueNRG public address (%d)", ret);
+			return ret;
```
But why? That just makes for more refactoring effort if anything else fallible is ever added to `bt_spi_bluenrg_setup()`, might as well do it properly from the start.

msmttchr replied:
Agreed. Thanks.

jhedberg commented:
> > * I would leave to the user to select whether he wants to enable the flexibility to add the public address (and include the relevant code) . Looking at [BlueNRG: allow setting public address #65179 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/65179#discussion_r1410570327), I guess you do not agree, but just to have your comment.
> 
> Same as point 1, I suppose, mostly a code size issue.

FWIW, I'm fine with how this is now (i.e. with the code size drawback). However, if you ever wanted to make it possible to have BlueNRG support without any setup callback, you'd need to add a new BlueNRG-specific Kconfig option for that (which would make the existing `select` statement conditional: `select BT_HCI_SETUP if BT_SPI_BLUENRG_PUB_ADDR`). But as I said, this seems slightly overkill right now.

Regarding the preprocessor conditions for including the setup callback, right now BlueNRG is the only "extension" of the SPI HCI driver, so I don't see any problem with this. If things get more complicated in the future and we start having multiple variants of the HCI SPI driver, then we should probably start considering splitting this up into multiple HCI drivers for simplicity. This would come with some risk of code duplication, but might be worth it, since otherwise you start essentially getting "drivers within a drivers" suff which can get quite messy and harder to maintain.

msmttchr commented:
> > 
> > I've tested on an stm32f767. Since this doesn't touch the low-level SPI protocol itself, I hope there's nothing platform-specific about this and it will work everywhere.
> 
> Ok, I will do some smoke test, but I think, we should be good now.

Smoke test done. It looks good. Thanks.

msmttchr reviewed (approved):
I tested briefly and it looks good to me.
I left minor comment about documentation update, but I would prefer to have this PR merged as soon as possible, so feel free to ignore it, if you do not agree.

arbrauns committed:
drivers: bluetooth: add more documentation for BT_HCI_SET_PUBLIC_ADDR

Hopefully this should now aid driver developers as well as application
developers.

Signed-off-by: Armin Brauns <armin.brauns@embedded-solutions.at>

msmttchr commented:
@jhedberg, if it is not a problem, would it be possible for you to approve and merge this PR ?
